torch/_higher_order_ops/scan.py

```diff
@@ -357,9 +456,176 @@ def scan_op_dense(combine_fn, init, xs, dim, reverse, additional_inputs):
+        operands_outs = ctx.saved_tensors
+        init = operands_outs[:num_leaves_init]
+        # The backward scan operates on the 0-th dim and thus the original inputs need to be
+        # permuted accordingly
+        xs = [
+            shift_source_dim_to_target_dim(o, dim, bwd_scan_dim)
+            for o in operands_outs[num_leaves_init : num_leaves_init + num_leaves_xs]
+        ]
+        # The forward path stores init + xs + carries + additional_inputs
+        # Thus carries (of length num_leaves_init) start at positions num_leaves_init + num_leaves_xs
+        # and end with 2 * num_leaves_init + num_leaves_xs
+        carries = operands_outs[
+            num_leaves_init + num_leaves_xs : 2 * num_leaves_init + num_leaves_xs
+        ]
+        additional_inputs = list(operands_outs[2 * num_leaves_init + num_leaves_xs :])
+
+        with torch._C._AutoDispatchBelowAutograd():
+            # The flat gradients are a list of g_c_T, g_ys and optionally the gradients for the additional_inputs
+            flat_grads_list = list(flat_grads)
+            g_c_T = flat_grads_list[:num_leaves_init]
+            g_ys = flat_grads_list[num_leaves_init : num_leaves_init + num_leaves_ys]
+            old_g_additional_inputs_T = flat_grads_list[
+                num_leaves_init + num_leaves_ys :
+            ]
+
+            # Prepare the inputs for the backward scan.
+            # This involves flipping the input xs if needed as well as
+            # Prepending the init of the forward scan to the carries
+            xs, carries = prepare_xs_carries_for_bwd(
+                xs, init, carries, bwd_scan_dim, reverse
+            )
+            xs_bwd = [*g_ys, *carries, *xs]
+
+            g_init, g_xs = _extract_carry_and_out(
+                scan_op(
+                    joint_graph, g_c_T, xs_bwd, bwd_scan_dim, True, additional_inputs
+                ),
+                num_leaves_init,
+            )
+
+            # g_xs may contain the gradients for the additional inputs as well which need to be separated
+            new_g_additional_inputs = g_xs[num_leaves_xs:] + old_g_additional_inputs_T
+            g_xs = g_xs[:num_leaves_xs]
+            g_xs = prepare_final_gradients_xs(g_xs, dim, reverse)
+
+        return *[None] * 6, *g_init, *g_xs, *new_g_additional_inputs
+
+
+@scan_op.py_impl(DispatchKey.Autograd)
+def scan_autograd(combine_fn, init, xs, dim, reverse, additional_inputs):
+    # A shortcut for the case where all inputs don't require gradient,
+    # we skip tracing the forward and backward graph.
+    if pytree.tree_all_only(
+        torch.Tensor,
+        lambda t: not t.requires_grad,  # type: ignore[union-attr]
+        (init, xs),
+    ):
+        with torch._C._AutoDispatchBelowAutograd():
+            return scan_op(combine_fn, init, xs, dim, reverse, additional_inputs)
+
+    num_leaves_init = len(init)
+    num_leaves_additional_inputs = len(additional_inputs)
+
+    if not torch.is_grad_enabled():
```

Comment:
This worths a comment on when this could happen: e.g. in the double-backward case when we trigger the autogradKey of backward. Can merge with the previous TODO.

probably use a different variable name like new_dim and add a comment on why we force it to be 0.

Reply:
Done